### PR TITLE
Webc 1658 implement offline detection and slate

### PR DESCRIPTION
**Pull request by eitanavgil**

### CODE REVIEW

File: src/plugin.tsx [hunk 1]
```diff
@@ -64,18 +71,17 @@ export class KalturaLivePlugin implements OnMediaUnload, OnRegisterUI, OnMediaLo
         this._player.addEventListener(this._player.Event.SOURCE_SELECTED, this._isEntryLiveType);
     }
 
+    onRegisterUI(uiManager: UIManager) {
+        this._overlayManager = this._contribServices.overlayManager;
```
Comment (esakal) [inline src/plugin.tsx]:
```suggestion
      // TODO remove once DI works correctly 
      const temp = this._contribServices.overlayManager;
```

File: src/plugin.tsx [hunk 2]
```diff
@@ -45,6 +48,10 @@ export class KalturaLivePlugin implements OnMediaUnload, OnRegisterUI, OnMediaLo
     // represents if the entry is in live mode or unknown
     private _broadcastState: LiveBroadcastStates = LiveBroadcastStates.Unknown;
 
+    private _isLiveApiCallTimeout: any = null;
+    private _overlayItem: OverlayItem | null = null;
+    private _overlayManager: OverlayManager | null = null;
```
Comment (esakal) [inline src/plugin.tsx]: can be removed

File: src/plugin.tsx [hunk 3]
```diff
@@ -93,22 +99,59 @@ export class KalturaLivePlugin implements OnMediaUnload, OnRegisterUI, OnMediaLo
...
             this.updateLiveStatus();
         }
     };
 
+    private _setLive = () => {
+        if (this._overlayItem) {
+            this._contribServices.overlayManager.remove(this._overlayItem);
+            this._overlayItem = null;
+        }
+        this._broadcastState = LiveBroadcastStates.Live;
```
Comment (esakal) [inline src/plugin.tsx]: move up

File: src/plugin.tsx [hunk 4]
```diff
@@ -118,6 +161,8 @@ export class KalturaLivePlugin implements OnMediaUnload, OnRegisterUI, OnMediaLo
                         method: "updateLiveStatus"
                     }
                 );
+                // re-check isLive on timeout
+                this._initTimeout();
             },
             error => {
                 if (error instanceof KalturaClientException) {
```
Comment (esakal) [inline src/plugin.tsx]: remove this condition, always show slate

File: src/plugin.tsx [hunk 5]
```diff
@@ -93,22 +99,59 @@ export class KalturaLivePlugin implements OnMediaUnload, OnRegisterUI, OnMediaLo
...
         const { pluginConfig } = this._configs;
         const protocol = KalturaPlaybackProtocol.hls;
         const id = this._player.config.sources.id; // todo - consider do this once on  media load
         const request = new LiveStreamIsLiveAction({ id, protocol });
         this._kalturaClient.request(request).then(
             data => {
                 if (data === true) {
-                    this._broadcastState = LiveBroadcastStates.Live;
+                    this._setLive();
                 }
```
Comment (esakal) [inline src/plugin.tsx]: add the handling of `data === false`

File: src/plugin.tsx [hunk 6]
```diff
@@ -129,18 +174,18 @@ export class KalturaLivePlugin implements OnMediaUnload, OnRegisterUI, OnMediaLo
                     });
                 } else if (error instanceof KalturaAPIException) {
                     // TODO - remove to the success part once TS client is fixed
```
Comment (esakal) [inline src/plugin.tsx]: remove // todo comment

File: src/middleware/live-middleware.ts [hunk 7]
```diff
@@ -1,19 +1,66 @@
...
+        } else {
+            // TODO - manage this (stop the timeout when the plugin / class is destroyed)
+            setTimeout(() => {
+                this.checkLiveStatus();
+            }, 250);
+        }
+    }
+
+    // sample the plugin if it live
+    private _checkIfLive() {
```
Comment (esakal) [inline src/middleware/live-middleware.ts]: -isPlayerLive

File: src/middleware/live-middleware.ts [hunk 8]
```diff
@@ -1,19 +1,66 @@
...
     constructor(plugin: KalturaLivePlugin) {
         super();
         this._livePlugin = plugin;
     }
 
+    // plugin does not have reference/direct API to the middleware - so we are sampling this.
+    // once we understand that we have a positive isLive - we clean and release the middleware
+    // if not - try again
+    checkLiveStatus() {
+        if (this._nextPlay && this._nextLoad && this._checkIfLive()) {
```
Comment (esakal) [inline src/middleware/live-middleware.ts]: missing initial check
```
if (!this._isFirstPlay) {
return;
}
```

File: src/middleware/live-middleware.ts [hunk 9]
```diff
@@ -1,19 +1,66 @@
...
 
     constructor(plugin: KalturaLivePlugin) {
         super();
         this._livePlugin = plugin;
     }
 
+    // plugin does not have reference/direct API to the middleware - so we are sampling this.
+    // once we understand that we have a positive isLive - we clean and release the middleware
+    // if not - try again
+    checkLiveStatus() {
```
Comment (esakal) [inline src/middleware/live-middleware.ts]: consider `initialPlayHandling()`

File: src/middleware/live-middleware.ts [hunk 10]
```diff
@@ -1,19 +1,66 @@
...
+    private _checkIfLive() {
+        return this._livePlugin.broadcastState === LiveBroadcastStates.Live;
+    }
+
     public load(next: Function): void {
-        super.callNext(next);
+        // if plugin is not active (E.G. in VOD) the middleware will not work
+        if (!this._livePlugin.isLiveEntry()) {
+            this.callNext(next);
+        }
```
Comment (esakal) [inline src/middleware/live-middleware.ts]: missing `return`

File: src/middleware/live-middleware.ts [hunk 11]
```diff
@@ -1,19 +1,66 @@
...
+        //we want to halt the load lifecycle method if we are in the 1st play or if we are not online
+        if (this._isFirstPlay || this._checkIfLive()) {
```
Comment (esakal) [inline src/middleware/live-middleware.ts]:
```suggestion
        if (this._checkIfLive()) {
```

File: src/middleware/live-middleware.ts [hunk 12]
```diff
@@ -1,19 +1,66 @@
...
+            this.callNext(next);
+        }
     }
 
     public play(next: Function): void {
-        super.callNext(next);
+        // if plugin is not active (E.G. in VOD) the middleware will not work
+        if (!this._livePlugin.isLiveEntry()) {
+            this.callNext(next);
+        }
```
Comment (esakal) [inline src/middleware/live-middleware.ts]: missing `return`

File: src/middleware/live-middleware.ts [hunk 13]
```diff
@@ -1,19 +1,66 @@
...
+        //Todo: Eitan - try simplifying or make more readable.
+        if (this._isFirstPlay || this._checkIfLive()) {
```
Comment (esakal) [inline src/middleware/live-middleware.ts]: should remove the ` || this._checkIfLive()` check
```
if (this._isFirstPlay) {
            this._nextPlay = next;
        } else {
            this.callNext(next);
        }
```

File: src/components/offline/offline.scss [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+    flex-direction: column;
+    align-items: center;
+    justify-content: center;
+    width: 100%;
+    height: 100%;
+    border-radius: 4px;
+    background-color: #000000;
+    .offline-icon {
+        width: 20vh;
+        height: 20vh;
```
Comment (eitanavgil) [inline src/components/offline/offline.scss]: Ask Dana how the UI looks on FullScreen

File: src/plugin.tsx [hunk 15]
```diff
@@ -93,22 +99,59 @@ export class KalturaLivePlugin implements OnMediaUnload, OnRegisterUI, OnMediaLo
 
     private _isEntryLiveType = () => {
         const { playerConfig } = this._configs;
+        // TODO S.M.- check if this is a safer way to check if this entry is a live-entry
+        // E.G. (this._player as any).isLive();
```
Comment (eitanavgil) [inline src/plugin.tsx]: Please apply and remove todos

File: src/plugin.tsx [hunk 16]
```diff
@@ -129,18 +174,18 @@ export class KalturaLivePlugin implements OnMediaUnload, OnRegisterUI, OnMediaLo
                     });
                 } else if (error instanceof KalturaAPIException) {
                     // TODO - remove to the success part once TS client is fixed
-                    this._broadcastState = LiveBroadcastStates.Offline;
+                    this._setOffline();
                     logger.error("Api exception", {
```
Comment (eitanavgil) [inline src/plugin.tsx]: Add meaningful log - e.g. 'Failed to call isLive API"

**Pull request merged**